### ISSUE TITLE
Cosmic freezes when one opens Firefox

### ISSUE THREAD

**Issue by garare992** [lilyinstarlight/nixos-cosmic#335]

**Comment by michaelBelsanti** [comment]
I'm having this issue on my AMD Framework Laptop, but not on my desktop that also uses AMD graphics.

**Comment by lilyinstarlight** [comment]
Can someone share the `journalctl` log from a boot where cosmic froze?

**Comment by garare992** [comment]
I'm also using an AMD Laptop. 
![image](https://github.com/user-attachments/assets/559e07ef-5d97-4f4f-ad78-a473c4f929fb)

**Comment by garare992** [comment]
@lilyinstarlight how exactly am i supposed to use journalctl?

**Comment by garare992** [comment]
More info that I found after some tinkering just now:

- Firefox only freezes/lags the cursor when it's maximized. I'm currently using Firefox in a non-maximized window with no issues. Yay!

- Right clicking anywhere on a maximized Firefox window unfreezes the mouse or makes it way less laggy, which allows you to quit the application or to make the window smaller, allowing you to use it.

**Comment by michaelBelsanti** [comment]
good info! I can also confirm that it seems to occur with most Firefox forks.

Here's my journalctl output as well: https://gist.github.com/michaelBelsanti/379566d70ac73d6b048c5d7d3fea5b9a
I've already taken a look through and didn't really find any useful information.

**Comment by michaelBelsanti** [comment]
this patch has fixed it for me https://github.com/pop-os/cosmic-comp/issues/887#issuecomment-2380399199

**Comment by garare992** [comment]
> this patch has fixed it for me [pop-os/cosmic-comp#887 (comment)](https://github.com/pop-os/cosmic-comp/issues/887#issuecomment-2380399199)

How do I apply that patch?

**Comment by lilyinstarlight** [comment]
> How do I apply that patch?

Something like this in your nixos config:

```diff
nixpkgs.overlays = [
  (final: prev: {
    cosmic-comp = prev.cosmic-comp.overrideAttrs (prevAttrs: {
      patches = (prevAttrs.patches or []) ++ [
        (pkgs.writeText "cosmic-comp-disable-direct-scanout.patch" ''
          diff --git a/src/backend/kms/surface/mod.rs b/src/backend/kms/surface/mod.rs
          index d0cfb8d..32aaf4a 100644
          --- a/src/backend/kms/surface/mod.rs
          +++ b/src/backend/kms/surface/mod.rs
          @@ -624,7 +624,8 @@ impl SurfaceThreadState {
                       cursor_size,
                       Some(gbm),
                   ) {
          -            Ok(compositor) => {
          +            Ok(mut compositor) => {
          +                compositor.use_direct_scanout(false);
                           self.active.store(true, Ordering::SeqCst);
                           self.compositor = Some(compositor);
                           Ok(())
        '')
      ];
    });
  })
];
```

**Comment by RingOfStorms** [comment]
I am not sure if this deserves a different issue or not. But for me I just get really bad performance when firefox is open, many small freezes. Nothing up to 30 seconds but up to about 10 seconds max at a time where I can't use the system at all.

cosmic-comp goes to the top of my cpu usage whenever trying to do anything in firefox. But in something like chrome everything is fine.

Just mousing the cursor over the firefox window (not maximized, or maximized, doesn't matter) causes the spikes and unresponsiveness.

I added the overlay above but that did not change anything for me:
https://git.joshuabell.xyz/dotfiles/~files/25e19e92567015bcf5724b4857c474f2ca89df34/hosts/gpdPocket3/configuration.nix?position=source-30.1-53.5-1

**Comment by lilyinstarlight** [comment]
@RingOfStorms Yeah that sounds unrelated. Your issue is more likely related to <https://github.com/pop-os/cosmic-comp/issues/749>/<https://github.com/pop-os/cosmic-comp/issues/703> if you are using client-side decorations in firefox

**Comment by RingOfStorms** [comment]
Thanks, it actually ended up being this one: I had no wallpaper/bg set...
https://github.com/pop-os/cosmic-comp/issues/638

It now works fine, wild that background can affect anything else like that

**Comment by lilyinstarlight** [comment]
Oh yeah that sounds like a bug... I was gonna fix default wallpapers very soon, it just needs some XDG_DATA_DIRS plumbing upstream

**Comment by lilyinstarlight** [comment]
For those using the above patch, please now use `environment.sessionVariables.COSMIC_DISABLE_DIRECT_SCANOUT = "1"` instead after updating this flake (ref: <https://github.com/pop-os/cosmic-comp/pull/925>)
